From 9e91d53956a73bb02471b086d991a60ae25eee10 Mon Sep 17 00:00:00 2001
From: Liu Shixin <liushixin2@huawei.com>
Date: Thu, 15 Jun 2023 17:56:38 +0800
Subject: [PATCH 029/780] [Huawei] mm/dynamic_hugetlb: fix race between
 hugetlb_pool_create() and hugetlb_pool_destroy()

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188903

--------------------------------

In the following scenarios:

CPU0                            CPU1

rmdir mem_cgroup                create dhugetlb_pool
  hugetlb_pool_destroy()
                                  hugetlb_pool_create()

hugetlb_pool_destroy() will do nothing since dhugetlb_pool has not been
created. But then hugetlb_pool_create() will create a new dhugetlb_pool,
the mem_cgroup and dhugetlb_pool will resieds in memory, and as interface
is deleted, we have no way to rmdir the mem_cgroup.

Since CSS_ONLINE flag will be cleard in cgroup_destroy process, use this
flag to determine whether to create dhugetlb_pool in hugetlb_pool_create().
Add a new global mutex lock dhugtlb_mutex to guarantee after
hugetlb_pool_destroy() and before hugetlb_pool_create(), CSS_ONLINE flag
is cleared. So hugetlb_pool_create() will skip creating new dhugetlb_pool.
And if hugetlb_pool_destroy() after hugetlb_pool_create(), dhugetlb_mutex
can guarantee the creation is completed before destruction.

Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 include/linux/dynamic_hugetlb.h |  5 +--
 kernel/cgroup/cgroup.c          |  5 ++-
 mm/dynamic_hugetlb.c            | 64 +++++++++++++++++++++++++--------
 3 files changed, 57 insertions(+), 17 deletions(-)

diff --git a/include/linux/dynamic_hugetlb.h b/include/linux/dynamic_hugetlb.h
index 5d6a2b77a5b1..3eb6003b2cf2 100644
--- a/include/linux/dynamic_hugetlb.h
+++ b/include/linux/dynamic_hugetlb.h
@@ -116,7 +116,8 @@ ssize_t write_hugepage_to_hpool(struct kernfs_open_file *of,
 				char *buf, size_t nbytes, loff_t off);
 int hugetlb_pool_info_show(struct seq_file *m, void *v);
 void hugetlb_pool_inherit(struct mem_cgroup *memcg, struct mem_cgroup *parent);
-int hugetlb_pool_destroy(struct cgroup *cgrp);
+int hugetlb_pool_destroy(struct cgroup *cgrp, bool *css_online);
+void hugetlb_pool_css_restore(struct cgroup *cgrp, bool css_online);
 void __init dynamic_hugetlb_init(void);
 
 struct page *alloc_page_from_dhugetlb_pool(gfp_t gfp, unsigned int order,
@@ -155,7 +156,7 @@ static inline bool dhugetlb_hide_files(struct cftype *cft)
 static inline void hugetlb_pool_inherit(struct mem_cgroup *memcg, struct mem_cgroup *parent)
 {
 }
-static inline int hugetlb_pool_destroy(struct cgroup *cgrp)
+static inline int hugetlb_pool_destroy(struct cgroup *cgrp, bool *css_online)
 {
 	return 0;
 }
diff --git a/kernel/cgroup/cgroup.c b/kernel/cgroup/cgroup.c
index bbf3930e4b7d..1574312b307d 100644
--- a/kernel/cgroup/cgroup.c
+++ b/kernel/cgroup/cgroup.c
@@ -5794,6 +5794,7 @@ static int cgroup_destroy_locked(struct cgroup *cgrp)
 	struct cgroup_subsys_state *css;
 	struct cgrp_cset_link *link;
 	int ssid;
+	bool css_online = false;
 
 	lockdep_assert_held(&cgroup_mutex);
 
@@ -5815,8 +5816,10 @@ static int cgroup_destroy_locked(struct cgroup *cgrp)
 	/*
 	 * If dynamic hugetlb is enabled, make sure dhugtlb_pool is free
 	 * before removing the corresponding memory cgroup.
+	 * If CSS_ONLINE, this function will clear it and set css_online
+	 * to true, if need, use hugetlb_pool_css_restore() to recover.
 	 */
-	if (hugetlb_pool_destroy(cgrp))
+	if (hugetlb_pool_destroy(cgrp, &css_online))
 		return -EBUSY;
 
 	/*
diff --git a/mm/dynamic_hugetlb.c b/mm/dynamic_hugetlb.c
index 2e244427bd5b..c43c2421f3ca 100644
--- a/mm/dynamic_hugetlb.c
+++ b/mm/dynamic_hugetlb.c
@@ -23,6 +23,7 @@ struct dhugetlb_pool_ops {
 
 bool enable_dhugetlb = false;
 DEFINE_STATIC_KEY_FALSE(dhugetlb_enabled_key);
+static DEFINE_MUTEX(dhugetlb_mutex);
 
 static bool enable_dpool;
 DEFINE_STATIC_KEY_FALSE(dpool_enabled_key);
@@ -987,18 +988,20 @@ static int hugetlb_pool_create(struct mem_cgroup *memcg, unsigned long nid,
 	return 0;
 }
 
-int hugetlb_pool_destroy(struct cgroup *cgrp)
+int hugetlb_pool_destroy(struct cgroup *cgrp, bool *css_online)
 {
 	struct cgroup_subsys_state *css = cgrp->subsys[memory_cgrp_id];
 	struct mem_cgroup *memcg = mem_cgroup_from_css(css);
-	struct dhugetlb_pool *hpool = memcg ? memcg->hpool : NULL;
+	struct dhugetlb_pool *hpool;
 	int ret = 0;
 
-	if (!dhugetlb_enabled)
+	if (!dhugetlb_enabled || !memcg)
 		return 0;
 
+	mutex_lock(&dhugetlb_mutex);
+	hpool = memcg->hpool;
 	if (!hpool || hpool->attach_memcg != memcg)
-		return 0;
+		goto unlock;
 
 	/*
 	 * Even if no process exists in the memory cgroup, some pages may still
@@ -1011,7 +1014,8 @@ int hugetlb_pool_destroy(struct cgroup *cgrp)
 		ret = hpool->ops->merge_pages(hpool);
 		if (ret) {
 			spin_unlock(&hpool->lock);
-			return -ENOMEM;
+			ret = -ENOMEM;
+			goto unlock;
 		}
 	}
 
@@ -1019,15 +1023,34 @@ int hugetlb_pool_destroy(struct cgroup *cgrp)
 		ret = hpool->ops->free_pool(hpool);
 		if (ret) {
 			spin_unlock(&hpool->lock);
-			return -ENOMEM;
+			ret = -ENOMEM;
+			goto unlock;
 		}
 	}
 	memcg->hpool = NULL;
 	spin_unlock(&hpool->lock);
 	put_hpool(hpool);
+unlock:
+	/*
+	 * Since hpool is destroyed and the memcg will be freed then,
+	 * clear CSS_ONLINE immediately to prevent race with create.
+	 */
+	if (!ret) {
+		if (cgrp->self.flags & CSS_ONLINE) {
+			cgrp->self.flags &= ~CSS_ONLINE;
+			*css_online = true;
+		}
+	}
+	mutex_unlock(&dhugetlb_mutex);
 	return ret;
 }
 
+void hugetlb_pool_css_restore(struct cgroup *cgrp, bool css_online)
+{
+	if (css_online)
+		cgrp->self.flags |= CSS_ONLINE;
+}
+
 static struct dhugetlb_pool_ops hugetlb_pool_ops = {
 	.free_pool = free_hugepage_to_hugetlb,
 	.merge_pages = hugetlb_pool_merge_all_pages,
@@ -1040,6 +1063,8 @@ static int hugetlb_pool_update(struct mem_cgroup *memcg,
 	bool new_create = false;
 	int ret = -EINVAL;
 
+	if (!(memcg->css.cgroup->self.flags & CSS_ONLINE))
+		return -EINVAL;
 again:
 	hpool = memcg->hpool;
 	if (!hpool) {
@@ -1214,7 +1239,9 @@ ssize_t write_hugepage_to_hpool(struct kernfs_open_file *of,
 	if (*endp != '\0' || size == 0)
 		return -EINVAL;
 
+	mutex_lock(&dhugetlb_mutex);
 	ret = hugetlb_pool_update(memcg, nid, size);
+	mutex_unlock(&dhugetlb_mutex);
 
 	return ret ? : nbytes;
 }
@@ -1426,19 +1453,29 @@ int dpool_init(struct dpool_info *arg)
 	struct dhugetlb_pool *hpool;
 	int i, ret, range_cnt;
 
+	mutex_lock(&dhugetlb_mutex);
 	if (!dpool_enabled || !arg || !arg->memcg || arg->memcg->hpool ||
-	    arg->range_cnt <= 0 || dpool_global_hpool)
-		return -EINVAL;
+	    arg->range_cnt <= 0 || dpool_global_hpool) {
+		ret = -EINVAL;
+		goto unlock;
+	}
+
+	if (!(arg->memcg->css.cgroup->self.flags & CSS_ONLINE)) {
+		ret = -EINVAL;
+		goto unlock;
+	}
 
 	ret = hugetlb_pool_create(arg->memcg, 0, &list_pool_ops);
 	if (ret)
-		return ret;
+		goto unlock;
 
 	hpool = arg->memcg->hpool;
 	range_cnt = arg->range_cnt;
 
-	if (!get_hpool_unless_zero(hpool))
-		return -EINVAL;
+	if (!get_hpool_unless_zero(hpool)) {
+		ret = -EINVAL;
+		goto unlock;
+	}
 
 	hpool->range_cnt = range_cnt;
 	hpool->pfn_ranges =
@@ -1451,7 +1488,6 @@ int dpool_init(struct dpool_info *arg)
 	memcpy(hpool->pfn_ranges, arg->pfn_ranges,
 	       sizeof(struct range) * range_cnt);
 
-	mutex_lock(&hpool->reserved_lock);
 	spin_lock(&hpool->lock);
 
 	for (i = 0; i < range_cnt; i++) {
@@ -1478,12 +1514,12 @@ int dpool_init(struct dpool_info *arg)
 	dpool_global_hpool = hpool;
 out:
 	spin_unlock(&hpool->lock);
-	mutex_unlock(&hpool->reserved_lock);
 put_pool:
 	if (ret)
 		put_hpool(hpool);
 	put_hpool(hpool);
-
+unlock:
+	mutex_unlock(&dhugetlb_mutex);
 	return ret;
 }
 
-- 
2.17.1

